### PR TITLE
MULTIARCH-4556: Introduce ImageStreamImportMode field in the image config

### PR DESCRIPTION
**Pull request by Prashanth684**
This is the api implementation for https://github.com/openshift/enhancements/pull/1605

- Introduces ImageStreamImportMode field in the image config spec and status - the status will be consumed by apiserver to set importMode for all imagestreams
- Introduce an `ImageStreamImportMode` tech preview feature gate
- Add feature gate test

### CODE REVIEW

**Comment by Prashanth684** [comment]
`verify` and `verify-crd-schema` are failing as these are changes to a crd which has not been touched in a while.

File: config/v1/types_image.go [hunk 1]
```diff
@@ -67,6 +68,14 @@ type ImageSpec struct {
 	// internal cluster registry.
 	// +optional
 	RegistrySources RegistrySources `json:"registrySources"`
+
+	// imageStreamImportMode controls the import mode behaviour of imagestreams.
+	// It can be set to `Legacy` or `PreserveOriginal`. If this value is specified,
```
Comment (JoelSpeed) [inline config/v1/types_image.go:73]: What do these two values do? What would an end user expect to see when they set one or the other?

File: config/v1/types_image.go [hunk 2]
```diff
@@ -67,6 +68,14 @@ type ImageSpec struct {
...
+	// this setting is applied to all newly created imagestreams which do not have the
+	// value set.
+	// +openshift:enable:FeatureGate=ImageStreamImportMode
+	// +optional
+	ImageStreamImportMode imagev1.ImportModeType `json:"imageStreamImportMode,omitempty"`
```
Comment (JoelSpeed) [inline config/v1/types_image.go:78]: Corss importing like this can cause issues, can we copy the type and enum values over please?

Comment (JoelSpeed) [inline config/v1/types_image.go:78]: Will need an enum tag to prevent invalid values

File: features/features.go [hunk 3]
```diff
@@ -548,4 +548,10 @@ var (
 							productScope(ocpSpecific).
 							enableIn(configv1.DevPreviewNoUpgrade, configv1.TechPreviewNoUpgrade).
 							mustRegister()
+	FeatureGateImageStreamImportMode = newFeatureGate("ImageStreamImportMode").
+                                                reportProblemsToJiraComponent("Multi-Arch").
+                                                contactPerson("psundara").
+                                                productScope(ocpSpecific).
+                                                enableIn(configv1.TechPreviewNoUpgrade).
```
Comment (JoelSpeed) [inline features/features.go:555]: You always need DevPreviewNoUpgrade if you have TechPreviewNoUpgrade

File: image/v1/tests/ImageStreamImportMode.yaml [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        apiVersion: config.openshift.io/v1
+        kind: Image
+        spec: {}
+      expected: |
+        apiVersion: config.openshift.io/v1
+        kind: Image
+        spec: {}
+        status:
+          internalRegistryHostname: image-registry.openshift-image-registry.svc:5000
+   - name: Should be able to configure imageStreamImportMode to Legacy
```
Comment (JoelSpeed) [inline image/v1/tests/ImageStreamImportMode.yaml:17]: What about testing for an invalid value?

File: image/v1/tests/ImageStreamImportMode.yaml [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        kind: Image
+        spec:
+          imageStreamImportMode: PreserveOriginal
+      expected: |
+        apiVersion: config.openshift.io/v1
+        kind: Image
+        spec:
+          imageStreamImportMode: PreserveOriginal
+        status:
+          internalRegistryHostname: image-registry.openshift-image-registry.svc:5000
```
Comment (JoelSpeed) [inline image/v1/tests/ImageStreamImportMode.yaml:43]: Is this a defaulted value?

File: config/v1/types_image.go [hunk 6]
```diff
@@ -1,6 +1,7 @@
 package v1
 
 import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+import imagev1 "github.com/openshift/api/image/v1"
```
Comment (deads2k) [inline config/v1/types_image.go:4]: this import shouldn't happen.  It couples our configuration to an image API that may evolve independently.

File: config/v1/types_image.go (hunk 1 above)
Comment (Prashanth684) [reply config/v1/types_image.go:73]: it sets the `importMode` field in an imagestream to control importing a single arch manifest vs the manifestlisted image: https://docs.openshift.com/container-platform/4.15/openshift_images/image-streams-manage.html#images-imagestream-specify-architecture_image-streams-managing

File: image/v1/tests/ImageStreamImportMode.yaml (hunk 5 above)
Comment (Prashanth684) [reply image/v1/tests/ImageStreamImportMode.yaml:43]: this value comes from the image-registry service by concatenating the <name>.<namespace>.svc:port and is the default for a cluster.

File: config/v1/types_image.go (hunk 6 above)
Comment (Prashanth684) [reply config/v1/types_image.go:4]: It does in a way depend on how the image API evolves though..if there is a change in the image API, I would assume we would want that to reflect in this field. Also, if we don't base this on the image API, then what about the field in the openshiftapiserver imagepolicyconfig? Would it be ok if that depends on the image API ? It is set based on the import mode in the image config status.

File: config/v1/types_image.go (hunk 2 above)
Comment (Prashanth684) [reply config/v1/types_image.go]: added.

Comment (Prashanth684) [reply config/v1/types_image.go]: done, but also had a clarification here: https://github.com/openshift/api/pull/1928#discussion_r1639304179

File: image/v1/tests/ImageStreamImportMode.yaml (hunk 4 above)
Comment (Prashanth684) [reply image/v1/tests/ImageStreamImportMode.yaml:17]: added.

File: features/features.go (hunk 3 above)
Comment (Prashanth684) [reply features/features.go]: done.

File: config/v1/types_image.go (hunk 1 above)
Comment (JoelSpeed) [reply config/v1/types_image.go:89]: Can you explain that in simple terms that don't need background knowledge, and then we can add that into the godoc? Something like `Legacy means ...`, `PreserveOriginal means ...`

File: image/v1/tests/ImageStreamImportMode.yaml (hunk 5 above)
Comment (JoelSpeed) [reply image/v1/tests/ImageStreamImportMode.yaml:43]: When writing the tests here, you must only add fields like this that are defaulted by the API itself. I don't think this one is API defaulted so I think this will break your tests presently

File: config/v1/types_image.go [hunk 7]
```diff
@@ -37,6 +37,23 @@ type Image struct {
 	Status ImageStatus `json:"status"`
 }
 
+// ImportModeType describes how to import an image manifest.
+// +enum
```
Comment (JoelSpeed) [inline config/v1/types_image.go:41]: What's this tag for? I don't recognise it

File: config/v1/types_image.go [hunk 8]
```diff
@@ -37,6 +37,23 @@ type Image struct {
...
+const (
+       // ImportModeLegacy indicates that the legacy behaviour should be used.
+       // For manifest lists, the legacy behaviour will discard the manifest list and import a single
+       // sub-manifest. In this case, the platform is chosen in the following order of priority:
+       // 1. tag annotations; 2. control plane arch/os; 3. linux/amd64; 4. the first manifest in the list.
+       // This mode is the default.
+       ImportModeLegacy ImportModeType = "Legacy"
+       // ImportModePreserveOriginal indicates that the original manifest will be preserved.
+       // For manifest lists, the manifest list and all its sub-manifests will be imported.
+       ImportModePreserveOriginal ImportModeType = "PreserveOriginal"
```
Comment (JoelSpeed) [inline config/v1/types_image.go:54]: This is good, we need to include this in the field description as well so end users will get this when they describe the field

File: config/v1/types_image.go [hunk 9]
```diff
@@ -67,6 +84,14 @@ type ImageSpec struct {
...
 	// +optional
 	RegistrySources RegistrySources `json:"registrySources"`
+
+	// imageStreamImportMode controls the import mode behaviour of imagestreams.
+	// It can be set to `Legacy` or `PreserveOriginal`. If this value is specified,
+	// this setting is applied to all newly created imagestreams which do not have the
+	// value set.
+	// +openshift:enable:FeatureGate=ImageStreamImportMode
+	// +optional
+	ImageStreamImportMode ImportModeType `json:"imageStreamImportMode,omitempty"`
```
Comment (JoelSpeed) [inline config/v1/types_image.go:94]: This is a configuration API, so we prefer to drop the omitempty, and allow `""` as an option for the enum.

That allows the rendered resource to show the field name and allows users to discover that this is a configurable option

File: config/v1/types_image.go [hunk 10]
```diff
@@ -83,6 +108,14 @@ type ImageStatus struct {
...
 	// +optional
 	ExternalRegistryHostnames []string `json:"externalRegistryHostnames,omitempty"`
+
+	// imageStreamImportMode controls the import mode behaviour of imagestreams. It can be
+	// `Legacy` or `PreserveOriginal` depending on the payload type of the cluster or the
+	// `ImageStreamImportMode` setting in the spec. This value is set by the image registry
+	// operator.
+	// +openshift:enable:FeatureGate=ImageStreamImportMode
+	// +optional
+	ImageStreamImportMode ImportModeType `json:"imageStreamImportMode,omitempty"`
```
Comment (JoelSpeed) [inline config/v1/types_image.go:118]: Can you explain the relationship between the spec and the status version of these fields?

File: hack/update-protobuf.sh [hunk 11]
```diff
@@ -10,7 +10,7 @@ To skip protobuf generation, set \$PROTO_OPTIONAL."
   exit 1
 fi
 
-if [[ "$(protoc --version)" != "libprotoc 23."* ]]; then
+if [[ "$(protoc --version)" != "libprotoc 3."* ]]; then
```
Comment (JoelSpeed) [inline hack/update-protobuf.sh:13]: Why are you moving this backwards?

File: config/v1/types_image.go (hunk 7 above)
Comment (Prashanth684) [reply config/v1/types_image.go:41]: which ? the enum?

File: config/v1/types_image.go (hunk 10 above)
Comment (Prashanth684) [reply config/v1/types_image.go:118]: the status will be reconciled based on:
1. the spec value 
2. if no spec value is specified, the image registry operator would look at the ClusterVersion status to determine the payload arch (TBD) and set the import mode accordingly, i.e single arch payload => the import mode is Legacy and multi payload => PreserveOriginal.

File: hack/update-protobuf.sh (hunk 11 above)
Comment (Prashanth684) [reply hack/update-protobuf.sh:13]: oops..i need to revert this..changed it based on my system

File: config/v1/types_image.go (hunk 9 above)
Comment (Prashanth684) [reply config/v1/types_image.go:94]: I did see that in some of the other enums..Does it mean it will always have a default value and be present in the spec? or does it mean configuring it empty is allowed?

File: config/v1/types_image.go (hunk 1 above)
Comment (Prashanth684) [reply config/v1/types_image.go:89]: i believe this is covered by adding the explanation to the fields themselves.

File: image/v1/tests/ImageStreamImportMode.yaml (hunk 5 above)
Comment (Prashanth684) [reply image/v1/tests/ImageStreamImportMode.yaml]: removed that field.

File: config/v1/types_image.go (hunk 9 above)
Comment (Prashanth684) [reply config/v1/types_image.go]: added the `""` enum option.

File: config/v1/types_image.go (hunk 7 above)
Comment (JoelSpeed) [reply config/v1/types_image.go:41]: Yep, I'd have expected something like `// +kubeuilder:validation:Enum:=A;B;C

File: config/v1/types_image.go (hunk 9 above)
Comment (JoelSpeed) [reply config/v1/types_image.go]: It means that by default it will show up as `imageStreamImportMode: ""`, which will show users that it is an option they could choose to configure. You just need to document what happens when it is the empty string

File: config/v1/types_image.go (hunk 10 above)
Comment (JoelSpeed) [reply config/v1/types_image.go:128]: I think those details should make it into the godoc for the field

File: config/v1/types_image.go (hunk 1 above)
Comment (JoelSpeed) [reply config/v1/types_image.go:89]: You need it in the fields, you've currently got it set only on the enum constants.

File: config/v1/types_image.go [hunk 12]
```diff
@@ -67,6 +84,19 @@ type ImageSpec struct {
...
+
+	// imageStreamImportMode controls the import mode behaviour of imagestreams.
+	// It can be set to `Legacy` or `PreserveOriginal`. If this value is specified,
+	// this setting is applied to all newly created imagestreams which do not have the
+	// value set. ImportModeLegacy indicates that the legacy behaviour should be used.
+	// For manifest lists, the legacy behaviour will discard the manifest list and import a single
+	// sub-manifest. In this case, the platform is chosen in the following order of priority:
+	// 1. tag annotations; 2. control plane arch/os; 3. linux/amd64; 4. the first manifest in the list.
+	// This mode is the default. ImportModePreserveOriginal indicates that the original manifest
+	// will be preserved. For manifest lists, the manifest list and all its sub-manifests will be imported.
```
Comment (JoelSpeed) [inline config/v1/types_image.go:96]:
```suggestion
	// imageStreamImportMode controls the import mode behaviour of imagestreams.
	// It can be set to `Legacy` or `PreserveOriginal` or the empty string. If this value is specified,
	// this setting is applied to all newly created ImageStreams which do not have the
	// value set. `Legacy` indicates that the legacy behaviour should be used.
	// For manifest lists, the legacy behaviour will discard the manifest list and import a single
	// sub-manifest. In this case, the platform is chosen in the following order of priority:
	// 1. tag annotations; 2. control plane arch/os; 3. linux/amd64; 4. the first manifest in the list.
	// `PreserveOriginal` indicates that the original manifest
	// will be preserved. For manifest lists, the manifest list and all its sub-manifests will be imported.
	// When empty, the `Legacy` behaviour is applied.
```
Just want to make this flow a bit with the language users will see. They care about the yaml naming, not the go naming

File: image/v1/tests/ImageStreamImportMode.yaml [hunk 13]
```diff
@@ -0,0 +1,47 @@
+apiVersion: apiextensions.k8s.io/v1 # Hack because controller-gen complains if we don't have this
+name: "Image"
+crdName: images.config.openshift.io
+tests:
```
Comment (JoelSpeed) [inline image/v1/tests/ImageStreamImportMode.yaml:4]: Needs the feature gate name in here as well

File: image/v1/tests/ImageStreamImportMode.yaml [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+   - name: Should be able to create a minimal Image config
+      initial: |
+        apiVersion: config.openshift.io/v1
+        kind: Image
+        spec: {}
+      expected: |
+        apiVersion: config.openshift.io/v1
+        kind: Image
+        spec: {}
+   - name: Should be able to configure imageStreamImportMode to Legacy
```
Comment (JoelSpeed) [inline image/v1/tests/ImageStreamImportMode.yaml:15]: Add a test for the empty string as well

File: openshiftcontrolplane/v1/types.go [hunk 15]
```diff
@@ -142,6 +143,13 @@ type ImagePolicyConfig struct {
 	// additionalTrustedCA is a path to a pem bundle file containing additional CAs that
 	// should be trusted during imagestream import.
 	AdditionalTrustedCA string `json:"additionalTrustedCA"`
+
+	// imageStreamImportMode provides the import mode value for  imagestreams.
+	// It can be `Legacy` or `PreserveOriginal`. If this value is specified,
+	// this setting is applied to all newly created imagestreams which do not have the
+	// value set.
+	// +optional
+	ImageStreamImportMode imagev1.ImportModeType `json:"imageStreamImportMode,omitempty"`
```
Comment (JoelSpeed) [inline openshiftcontrolplane/v1/types.go:152]: Need to copy the types over, do not import across the packages

File: image/v1/tests/ImageStreamImportMode.yaml (hunk 14 above)
Comment (Prashanth684) [reply image/v1/tests/ImageStreamImportMode.yaml:15]: i'm wondering how we can do this as the idea is that - when there is no import mode specified, it will be inferred based on the payload type and will be set by the image registry operator. When/where would these tests run?

File: config/v1/types_image.go (hunk 7 above)
Comment (Prashanth684) [reply config/v1/types_image.go:41]: remove this enum tag removes the enum listing from the [openapi](https://github.com/openshift/api/pull/1928/files#diff-6466e6ec9d299ee55a89f9ae0f455bc960d7295ead963490f5497447c2f1feb3R7224) files. is that ok ?

File: config/v1/types_image.go (hunk 12 above)
Comment (JoelSpeed) [reply config/v1/types_image.go:96]: Right ok, can you update my suggestion then with how you would word that? Probably something along the lines of
```
// When empty, the mode is determined based on the payload image specified in the ClusterVersion resource. 
```

File: image/v1/tests/ImageStreamImportMode.yaml (hunk 14 above)
Comment (JoelSpeed) [reply image/v1/tests/ImageStreamImportMode.yaml:15]: Well this is just an API integration test, so, you'd test that it is accepted as empty. Whatever mutation mechanism you choose to implement would be separate from these tests anyway. That test will likely want to be in origin.

When you say set by the operator, are you planning to set the spec? Typically that's not done and you re-calculate. If it's status, you can set it.

File: config/v1/types_image.go (hunk 7 above)
Comment (JoelSpeed) [reply config/v1/types_image.go:41]: Hmm, lets keep it. Might need to look at this tag again, I'll add a note to self

File: image/v1/tests/ImageStreamImportMode.yaml (hunk 14 above)
Comment (Prashanth684) [reply image/v1/tests/ImageStreamImportMode.yaml:15]: It is the status. If there is no value specified in the spec or it is "", the status would be updated to `Legacy` or `PreserveOriginal` based on the payload type that CVO advertises (single vs multi payload)

File: config/v1/types_image.go (hunk 10 above)
Comment (Prashanth684) [reply config/v1/types_image.go:131]: added.

File: config/v1/types_image.go (hunk 12 above)
Comment (Prashanth684) [reply config/v1/types_image.go]: made the appropriate changes.

File: image/v1/tests/ImageStreamImportMode.yaml (hunk 14 above)
Comment (Prashanth684) [reply image/v1/tests/ImageStreamImportMode.yaml:16]: > Whatever mutation mechanism you choose to implement would be separate from these tests anyway

So iiuc, this would just test the API standalone? not like an e2e?

Comment (Prashanth684) [reply image/v1/tests/ImageStreamImportMode.yaml:16]: added the test for empty string.

File: openshiftcontrolplane/v1/types.go [hunk 16]
```diff
@@ -117,6 +117,21 @@ type RoutingConfig struct {
 	Subdomain string `json:"subdomain"`
 }
 
+// ImportModeType describes how to import an image manifest.
+type ImportModeType string
```
Comment (JoelSpeed) [inline openshiftcontrolplane/v1/types.go:121]: This needs the enum tags too

File: openshiftcontrolplane/v1/types.go [hunk 17]
```diff
@@ -142,6 +157,18 @@ type ImagePolicyConfig struct {
...
+	// It can be `Legacy` or `PreserveOriginal`. `Legacy` indicates that the legacy behaviour
+	// should be used. For manifest lists, the legacy behaviour will discard the manifest list
+	// and import a single sub-manifest. In this case, the platform is chosen in the following
+	// order of priority: 1. tag annotations; 2. control plane arch/os; 3. linux/amd64; 4. the first
+	// manifest in the list. `PreserveOriginal` indicates that the original manifest will be preserved.
+	// For manifest lists, the manifest list and all its sub-manifests will be imported.If this value
+	// is specified, this setting is applied to all newly created imagestreams which do not have the
+	// value set.
+	// +optional
+	ImageStreamImportMode ImportModeType `json:"imageStreamImportMode"`
```
Comment (JoelSpeed) [inline openshiftcontrolplane/v1/types.go:171]: Can't see any tests for this field. Also, should this field also be gated behind the featuregate?

Comment (Prashanth684) [reply openshiftcontrolplane/v1/types.go:171]: @JoelSpeed apologies again for my limited knowledge on this - this is not an API that can be directly controlled by the user, but it still needs to be behind a feature gate. This value is set by the openshift-apiserver once it reads the image config status.

File: openshiftcontrolplane/v1/types.go (hunk 16 above)
Comment (Prashanth684) [reply openshiftcontrolplane/v1/types.go:121]: i can add it - but i was unsure because these fields are set by apiserver based on the different crd statuses and not controlled directly.

File: openshiftcontrolplane/v1/types.go (hunk 17 above)
Comment (JoelSpeed) [reply openshiftcontrolplane/v1/types.go:171]: If you don't have it gated, you've effectively GA'd this straight away, at which point we can't remove it. Since the other field is gated, I'd expect them both to be gated, and to have them promoted at the same time

File: openshiftcontrolplane/v1/types.go (hunk 16 above)
Comment (JoelSpeed) [reply openshiftcontrolplane/v1/types.go:121]: We should have the validation even if they are set by controllers

File: openshiftcontrolplane/v1/types.go (hunk 17 above)
Comment (Prashanth684) [reply openshiftcontrolplane/v1/types.go:174]: > Can't see any tests for this field

afaik the fields here are synced by the openshift-apiserver-operator in the observed config from the value of the status fields in the various other config CRs. This value is meant to be generated the same way - by looking at the `image.config.openshift.io` status and putting that in the observedconfig.

Comment (Prashanth684) [reply openshiftcontrolplane/v1/types.go:174]: @JoelSpeed about my above question - are tests needed for this? i believe this is the only open item left to resolve.

Comment (Prashanth684) [reply openshiftcontrolplane/v1/types.go:174]: The `OpenShiftAPIServer` api is where this value is reflected in the observed config. There is a basic test for that API [here](https://github.com/openshift/api/blob/0d46442e8df17a87cfcf6666ab19b28e88620b59/operator/v1/tests/openshiftapiservers.operator.openshift.io/AAA_ungated.yaml). I am not sure how we can add a test for an object which is generated by the apiserver.

Comment (Prashanth684) [reply openshiftcontrolplane/v1/types.go:174]: I also think the `OpenshiftAPIServerConfig` was used in v3.11, but was deprecated since.

Comment (JoelSpeed) [reply openshiftcontrolplane/v1/types.go:174]: There's an explanation of how those style of tests work in https://github.com/openshift/api/tree/master/tests, but yes, I would expect some basic testing even now.

The openshift-apiserver isn't the only implementation of these APIs, other topologies of OpenShift use the CRDs, so we should make sure the validations work on the CRDs too

Comment (Prashanth684) [reply openshiftcontrolplane/v1/types.go:174]: thanks for the pointers. added a test for the featuregate.

Comment (Prashanth684) [reply openshiftcontrolplane/v1/types.go:174]: @JoelSpeed is the test file that i added sufficient?

**Comment by JoelSpeed** [comment]
There appears to be a yaml error in the integration tests you've added, check the failures

File: operator/v1/tests/openshiftapiservers.operator.openshift.io/ImageStreamImportMode.yaml [hunk 18]
```diff
@@ -0,0 +1,57 @@
+apiVersion: apiextensions.k8s.io/v1 # Hack because controller-gen complains if we don't have this
+name: "OpenShiftAPIServer"
+crdName: openshiftapiservers.operator.openshift.io
+featureGate: ImageStreamImportMode
+tests:
+  onCreate:
+    - name: Should be able to create a minimal OpenShiftAPIServer with import mode set to Legacy
```
Comment (JoelSpeed) [inline operator/v1/tests/openshiftapiservers.operator.openshift.io/ImageStreamImportMode.yaml:7]: Can you please add a test case that shows an invalid value being rejected

**Comment by JoelSpeed** [comment]
> could not load schema check generation contexts from dir "/go/src/github.com/openshift/api/image/v1/tests": could not unmarshal YAML for type meta inspection: error converting YAML to JSON: yaml: line 7: mapping values are not allowed in this context

Still something wrong in the test files, from verify-crd-schema

**Comment by JoelSpeed** [comment]
Please can you resolve the following
```
error in config/v1/zz_generated.crd-manifests/0000_10_config-operator_01_images-CustomNoUpgrade.crd.yaml: ListsMustHaveSSATags: crd/images.config.openshift.io version/v1 field/^.spec.externalRegistryHostnames must set x-kubernetes-list-type
		error in config/v1/zz_generated.crd-manifests/0000_10_config-operator_01_images-CustomNoUpgrade.crd.yaml: ListsMustHaveSSATags: crd/images.config.openshift.io version/v1 field/^.spec.registrySources.allowedRegistries must set x-kubernetes-list-type
		error in config/v1/zz_generated.crd-manifests/0000_10_config-operator_01_images-CustomNoUpgrade.crd.yaml: ListsMustHaveSSATags: crd/images.config.openshift.io version/v1 field/^.spec.registrySources.blockedRegistries must set x-kubernetes-list-type
		error in config/v1/zz_generated.crd-manifests/0000_10_config-operator_01_images-CustomNoUpgrade.crd.yaml: ListsMustHaveSSATags: crd/images.config.openshift.io version/v1 field/^.spec.registrySources.insecureRegistries must set x-kubernetes-list-type
		error in config/v1/zz_generated.crd-manifests/0000_10_config-operator_01_images-CustomNoUpgrade.crd.yaml: ListsMustHaveSSATags: crd/images.config.openshift.io version/v1 field/^.spec.allowedRegistriesForImport must set x-kubernetes-list-type
		error in config/v1/zz_generated.crd-manifests/0000_10_config-operator_01_images-CustomNoUpgrade.crd.yaml: ListsMustHaveSSATags: crd/images.config.openshift.io version/v1 field/^.status.externalRegistryHostnames must set x-kubernetes-list-type
```

SSA tags should be fixed as they come up

**Comment by Prashanth684** [comment]
> Please can you resolve the following
> 
> ```
> error in config/v1/zz_generated.crd-manifests/0000_10_config-operator_01_images-CustomNoUpgrade.crd.yaml: ListsMustHaveSSATags: crd/images.config.openshift.io version/v1 field/^.spec.externalRegistryHostnames must set x-kubernetes-list-type
> 		error in config/v1/zz_generated.crd-manifests/0000_10_config-operator_01_images-CustomNoUpgrade.crd.yaml: ListsMustHaveSSATags: crd/images.config.openshift.io version/v1 field/^.spec.registrySources.allowedRegistries must set x-kubernetes-list-type
> 		error in config/v1/zz_generated.crd-manifests/0000_10_config-operator_01_images-CustomNoUpgrade.crd.yaml: ListsMustHaveSSATags: crd/images.config.openshift.io version/v1 field/^.spec.registrySources.blockedRegistries must set x-kubernetes-list-type
> 		error in config/v1/zz_generated.crd-manifests/0000_10_config-operator_01_images-CustomNoUpgrade.crd.yaml: ListsMustHaveSSATags: crd/images.config.openshift.io version/v1 field/^.spec.registrySources.insecureRegistries must set x-kubernetes-list-type
> 		error in config/v1/zz_generated.crd-manifests/0000_10_config-operator_01_images-CustomNoUpgrade.crd.yaml: ListsMustHaveSSATags: crd/images.config.openshift.io version/v1 field/^.spec.allowedRegistriesForImport must set x-kubernetes-list-type
> 		error in config/v1/zz_generated.crd-manifests/0000_10_config-operator_01_images-CustomNoUpgrade.crd.yaml: ListsMustHaveSSATags: crd/images.config.openshift.io version/v1 field/^.status.externalRegistryHostnames must set x-kubernetes-list-type
> ```
> 
> SSA tags should be fixed as they come up

These are already existing fields in an already existing CRD. is the change simply to add `// +listType=atomic` ? Hopefully it won't impact the existing crd?

**Comment by JoelSpeed** [comment]
/override ci/prow/e2e-upgrade-minor

The minor upgrade has been broken since branching, I'm working on fixing it. Looking at the changes here the only upgrade potential issue would be the new SSA tags, but they are all atomic which is the default anyway, so there's no actual functional change
